broker-util/oo-admin-repair

```diff
@@ -403,6 +418,37 @@ def find_unresponsive_servers(available_servers, auto_confirm)
   unresponsive_servers
 end
 
+def find_orphaned_envs(namespace=nil)
+  orphaned_envs = []
+  domains = []
+  if namespace
+    begin
+      domains = [Domain.find_by(:namespace => namespace)]
+    rescue Mongoid::Errors::DocumentNotFound
+      print_message "Unable to find domain. The domain #{namespace} does not exist. Skipping..."
+    rescue Exception => e
+      print_message "Unable to find domain #{namespace} with error: #{e.message}."
+    end
+    # Find all component_ids for every application in this domain
+    component_ids = domains[0].applications.map {|a| a.component_instances.map {|ci| ci._id}}.flatten
+  else
+    domains = Domain.all
+    # find all component_ids for every application
+    component_ids = Application.all.map {|a| a.component_instances.map {|ci| ci._id}}.flatten
```

Comment:
Constructing objects for all domains or apps is definitely not ideal in OpenShift Online.  We _do_ want to be able to validate all domains, rather than waiting for people to report a problem, though.   For one data point, I ran this:

db.domains.find({ 'env_vars.component_id': { $exists: true }  }).count()

without divulging numbers, I can say that this result represents 0.3% of our domains.  That's still a large number, but perhaps manageable if you only pull the ids.  I kicked around mongo commands for doing this.  I'm pretty unskilled at mongo, but I managed this:

```
cids = db.domains.aggregate({ $project: { cid: '$env_vars.component_id' } }, { $unwind: "$cid" }, { $group: { _id: 'cid', res: {$addToSet:'$cid'} } }); true
cids2 = db.applications.aggregate({ $match: { 'component_instances._id': { $in: res } } }, { $project: { cid: '$component_instances._id' } }, { $unwind: "$cid" }, { $group: { _id: 'cid', res: {$addToSet:'$cid'} } }); true
valid_ids = []
cids2.result[0].res.forEach(function (key) {  valid_ids.push(key.str); })
orphan_ids = []
res.forEach(function (key) { if (valid_ids.indexOf(key.str) == -1) { orphan_ids.push(key); } })
```

The two 'aggregate' commands are the important mongo part.  The remaining lines are really just comparing object ids in two lists, and not very efficiently.  it would be better done using set difference in ruby.

But just to validate the correctness of the list:
```
libra:PRIMARY> orphan_ids.length
121
libra:PRIMARY> db.applications.find({ 'component_instances._id': { $in: orphan_ids } })
libra:PRIMARY> db.domains.find({ 'env_vars.component_id': { $in: orphan_ids } }).count()
120
```

so that all makes sense.  So I think this is the right approach:

1) find domains by those that have env vars with a component id, and pull out the component id.
2) find only applications matching the component ids referenced
3) subtract the ids found in step 2 from the ids found in step 1
4) what's left are your orphans.  Iterate over the domains containing those component ids and fix them.